File: src/test/java/com/ealanta/productapp/label/LabelledProductSourceTest.java
package com.ealanta.productapp.label;

import com.ealanta.productapp.misc.TestUtils;
import com.ealanta.productapp.price.PriceInfo;
import com.ealanta.productapp.price.PriceReduction;
import com.ealanta.productapp.product.Product;
import com.ealanta.productapp.product.ProductComparator;
import com.ealanta.productapp.service.ProductSource;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.*;
import org.mockito.junit.MockitoJUnitRunner;

import static org.mockito.Mockito.*;

import java.util.*;
import java.util.List;

@RunWith(MockitoJUnitRunner.class)
public class LabelledProductSourceTest {

    @Mock
    private ProductSource mProductSource;

    @Mock
    PriceLabelGenerator mLabelGenerator;

    private LabelledProductSource sut;

    private List<Product> productList;
    private Product p1;
    private Product p2;
    private Product p3;
    private PriceReduction pr1;
    private PriceReduction pr2;
    private PriceReduction pr3;

    @Before
    public void setup() {
         sut = new LabelledProductSource(new ProductComparator(),mProductSource,mLabelGenerator);
         p1 = getProduct("p1");
         p2 = getProduct("p2");
         p3 = getProduct("p3");
         productList = Arrays.asList(p1,p2,p3);
         pr1 = p1.getPriceInfo().getPriceReduction().get();
         pr2 = p2.getPriceInfo().getPriceReduction().get();
         pr3 = p3.getPriceInfo().getPriceReduction().get();
    }

    @Test
    public void testShowWasNow(){
        checkGetProducts(LabelType.ShowWasNow);
    }

    @Test
    public void testShowWasThenNow(){
        checkGetProducts(LabelType.ShowWasThenNow);
    }

    @Test
    public void testGetProducts(){
        checkGetProducts(LabelType.ShowPercDiscount);
    }

    public void checkGetProducts(LabelType labelType){

        when(mProductSource.getFilteredAndSortedProducts(
                any(),
                any())).thenReturn(productList);

        ArgumentCaptor<PriceReduction> prArg = ArgumentCaptor.forClass(PriceReduction.class);
        ArgumentCaptor<LabelType> lArg = ArgumentCaptor.forClass(LabelType.class);

        when(mLabelGenerator.generatePriceLabel(prArg.capture(),lArg.capture()))
                .thenReturn("LABEL1")
                .thenReturn("LABEL2")
                .thenReturn("LABEL3");

        List<LabelledProduct> result = sut.getProducts(labelType);

        for(LabelledProduct lp : result){
            System.out.printf("%s -> %s%n",lp.getProductId(),lp.getPriceLabel());
        }

        Assert.assertEquals(prArg.getAllValues(),Arrays.asList(pr1,pr2,pr3));
        Assert.assertEquals(3,lArg.getAllValues().size());
        Assert.assertTrue(lArg.getAllValues().stream().allMatch(p -> p.equals(labelType)));

        LabelledProduct lp1 = result.get(0);
        LabelledProduct lp2 = result.get(1);
        LabelledProduct lp3 = result.get(2);

        Assert.assertEquals("p1", lp1.getProductId());
        Assert.assertEquals("p2", lp2.getProductId());
        Assert.assertEquals("p3", lp3.getProductId());

        Assert.assertEquals("LABEL1", lp1.getPriceLabel());
        Assert.assertEquals("LABEL2", lp2.getPriceLabel());
        Assert.assertEquals("LABEL3", lp3.getPriceLabel());

        verify(mProductSource, times(1)).getFilteredAndSortedProducts(any(),any());
        verify(mLabelGenerator, times(3)).generatePriceLabel(any(),any());
        verifyNoMoreInteractions(mLabelGenerator,mProductSource);
    }

    private Product getProduct(String productId){
        Product result = new Product();
        PriceInfo info = new PriceInfo();
        PriceReduction pr = new PriceReduction(
                TestUtils.getMoney("100"),
                TestUtils.getMoney("90"));
        info.setPriceReduction(Optional.of(pr));
        result.setPriceInfo(info);
        result.setProductId(productId);
        return result;
    }
}

File: src/test/kotlin/com/ealanta/productapp/label/LabelledProductSourceTest.kt
package com.ealanta.productapp.label;

import com.ealanta.productapp.domain.LabelType
import com.ealanta.productapp.domain.LabelledProduct
import com.ealanta.productapp.domain.PriceInfo
import com.ealanta.productapp.domain.PriceReduction
import com.ealanta.productapp.domain.Product
import com.ealanta.productapp.utils.TestUtils

import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.junit.MockitoJUnitRunner
import java.util.*
import kotlin.Comparator
import kotlin.properties.Delegates
import com.nhaarman.mockito_kotlin.*
import kotlin.test.*

//@RunWith(MockitoJUnitRunner::class)
class LabelledProductSourceTest {

    /*
    val mProductSource: ProductSource = mock()

    val mLabelGenerator: PriceLabelFactory = mock()

    val prArg: KArgumentCaptor<PriceReduction> = argumentCaptor();

    val lArg: KArgumentCaptor<LabelType> = argumentCaptor();

    val comparator:Comparator<Product> = compareBy { it.productId }

    var sut: LabelledProductSource by Delegates.notNull()
    var productList: MutableList<Product> by Delegates.notNull()
    var p1: Product  by Delegates.notNull()
    var p2: Product  by Delegates.notNull()
    var p3: Product  by Delegates.notNull()
    var pr1: PriceReduction  by Delegates.notNull()
    var pr2: PriceReduction  by Delegates.notNull()
    var pr3: PriceReduction  by Delegates.notNull()

    @Before
    fun setup() {
        sut = LabelledProductSource( comparator, mProductSource, mLabelGenerator);
        p1 = getProduct("p1");
        p2 = getProduct("p2");
        p3 = getProduct("p3");
        productList = mutableListOf<Product>(p1, p2, p3);
        pr1 = p1.priceInfo!!.priceReduction!!;
        pr2 = p2.priceInfo!!.priceReduction!!;
        pr3 = p3.priceInfo!!.priceReduction!!;
    }

    @Test
    fun testShowWasNow() {
        checkGetProducts(LabelType.ShowWasNow);
    }

    @Test
    fun testShowWasThenNow() {
        checkGetProducts(LabelType.ShowWasThenNow);
    }

    @Test
    fun testGetProducts() {
        checkGetProducts(LabelType.ShowPercDiscount);
    }

    fun checkGetProducts(labelType: LabelType) {

        whenever(mProductSource.getFilteredAndSortedProducts(
                TestUtils.any(),
                TestUtils.any())).thenReturn(productList);


        whenever(mLabelGenerator.generatePriceLabel(prArg.capture(), lArg.capture()))
                .thenReturn("LABEL1")
                .thenReturn("LABEL2")
                .thenReturn("LABEL3");

        val result = sut.getProducts(labelType);

        result.forEach {
            System.out.printf("%s -> %s%n", it.productId, it.priceLabel);
        }

        assertEquals(prArg.allValues, Arrays.asList(pr1, pr2, pr3));
        assertEquals(3, lArg.allValues.size);
        assertTrue(lArg.allValues.all { it.equals(labelType) })

        val lp1: LabelledProduct = result[0]
        val lp2: LabelledProduct = result[1]
        val lp3: LabelledProduct = result[2]

        assertEquals("p1", lp1.productId);
        assertEquals("p2", lp2.productId);
        assertEquals("p3", lp3.productId);

        assertEquals("LABEL1", lp1.priceLabel);
        assertEquals("LABEL2", lp2.priceLabel);
        assertEquals("LABEL3", lp3.priceLabel);

        verify(mProductSource, times(1)).getFilteredAndSortedProducts(TestUtils.any(), TestUtils.any());
        verify(mLabelGenerator, times(3)).generatePriceLabel(TestUtils.any(), TestUtils.any());
        verifyNoMoreInteractions(mLabelGenerator, mProductSource);
    }

    fun getProduct(productId: String): Product {
        val result = Product();
        val info = PriceInfo();
        val pr = PriceReduction(
                TestUtils.getMoney("100")!!, TestUtils.getMoney("90")!!);
        info.priceReduction = pr
        result.priceInfo = info;
        result.productId = productId;
        return result;
    }

     */
}
